From 8c27da1ec313de1e316ab15512f6ee0c102e5fc0 Mon Sep 17 00:00:00 2001
From: "Radivoje (Ogi) Jovanovic" <rjovanovic@nvidia.com>
Date: Thu, 5 Jan 2023 17:23:01 -0800
Subject: [PATCH] drivers: i2c: add timeout to arbitrator

if arbitration is held by userspace for more than TIMEOUT seconds
give away arbitration.

Signed-off-by: Radivoje (Ogi) Jovanovic <rjovanovic@nvidia.com>
Upstream-Status: Pending [Not submitted to upstream yet]
---
 drivers/i2c/muxes/i2c-arb-gpio-challenge.c | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
index 69da27657062..5993070028fb 100644
--- a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
+++ b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
@@ -18,6 +18,7 @@
 #include <linux/device.h>
 #include <linux/workqueue.h>
 #include <linux/spinlock.h>
+#define BUS_TIMEOUT 600
 #endif //CONFIG_DRAGON_CHASSIS
 
 /**
@@ -45,6 +46,7 @@ struct i2c_arbitrator_data {
 	struct workqueue_struct *wq;
 	struct delayed_work dwork;
 	spinlock_t bus_locked_spin;
+	int cnt;
 #endif //CONFIG_DRAGON_CHASSIS
 };
 
@@ -202,6 +204,7 @@ static ssize_t arb_state_store(struct device *dev,
 			return ret;
 		spin_lock(&arb->bus_locked_spin);
 		arb->bus_locked = true;
+		arb->cnt = 0;
 		spin_unlock(&arb->bus_locked_spin);
 		queue_delayed_work(arb->wq, &arb->dwork, 0);
 	}
@@ -249,8 +252,20 @@ static void pet_cpld_register(struct work_struct *work)
 			i2c_arbitrator_deselect(muxc, 0);
 
 		}
-		else
-			queue_delayed_work(arb->wq, &arb->dwork, msecs_to_jiffies(1000));
+		else {
+			spin_lock(&arb->bus_locked_spin);
+			arb->cnt++;
+			spin_unlock(&arb->bus_locked_spin);
+			if (arb->cnt >= BUS_TIMEOUT) {
+				dev_err(muxc->dev, "Bus was held for more than %d seconds. Cancel arbitration\n", BUS_TIMEOUT);
+				spin_lock(&arb->bus_locked_spin);
+				arb->bus_locked = false;
+				spin_unlock(&arb->bus_locked_spin);
+				i2c_arbitrator_deselect(muxc, 0);
+			}
+			else
+				queue_delayed_work(arb->wq, &arb->dwork, msecs_to_jiffies(1000));
+		}
 	}
 	else
 	{
-- 
2.43.0

